### ISSUE TITLE
`GetConstantValue` should used boxed integers instead of boxing every time

### ISSUE THREAD

**Issue by Youssef1313** [dotnet/runtime#90718]
### Description

![image](https://github.com/dotnet/runtime/assets/31348972/c485edc3-3904-4cea-b869-a805325e7084)

### Reproduction Steps

-

### Expected behavior

Shouldn't be unnecessary boxing

### Actual behavior

This is boxing.

https://github.com/dotnet/runtime/blob/ce54acdc1c2e820f06d6f813d6b6ed6ce0a1b64d/src/tools/illink/src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs#L363

**Comment by ShreyasJejurkar** [comment]
Not sure how a boxed integer would help, as it might return `string` as well! 

https://github.com/dotnet/runtime/blob/ce54acdc1c2e820f06d6f813d6b6ed6ce0a1b64d/src/tools/illink/src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs#L396-L407

**Comment by Youssef1313** [comment]
string can remain. There is no boxing concern with it.

**Comment by EgorBo** [comment]
If boxing is indeed an issue, feel free to just define `static readonly` fields representing boxed values and use those.

**Comment by vitek-karas** [comment]
I don't think we can use static fields since the values are not known up front - they're read from the IL stream.
If this is indeed a perf problem, then we could probably avoid the cast to `(int)` and just return the `instruction.Operand` which will be an already boxed value. The casts are really not doing much here.

**Comment by EgorBo** [comment]
Even for unknowns, if it makes sense, you can define an array of, say, 200 objects representing -100..100 (at least for the `Ldc_I4` which is, I guess, the most popular).

A few big projects use a similar trick, e.g.: https://grep.app/search?q=static%20readonly%20object%20Box

**Comment by vitek-karas** [comment]
Good points - thanks a lot!

I agree it would make sense to use precreated boxed values for the constants and avoid casting for the cases where we read the value from the operand (as that will already be boxed).

**Comment by vitek-karas** [comment]
@Youssef1313 it's really cool you're looking into performance of illink. We do have a benchmark for it: https://github.com/dotnet/performance/tree/main/src/benchmarks/real-world/ILLink, so it would be really great if you could attach potential improvements in the benchmark to each of the changes.

That said, the benchmark probably needs some work:
* No good way to run using a local build of illink
* No good way to compare local build to official build
* The benchmarked app is probably too simplistic (HelloWorld)

But if we're going to do perf work for illink, it would make a lot of sense to invest into testing of it as well.
